Add a dependency injection container

markstory opened the pull request:
This branch will collect the various incremental changes as support for a dependency injection container are added. The plan for this work can be found in #14865

markstory committed:
Add Core/ContainerInterface

Add a DI container interface to give us an interface that applications
and plugins can use when registering services. The PSR interfaces do not
provide methods to register services. However, for us to have a useful
container interface we need those methods as plugins and applications
will want a stable interface. Should we need to change off of
league/container this interface is small enough that we could adapt
it to a new container implementation. This interface will also allow
ambitious applications to use an alternative container implementation
and not break their plugins.

Refs #14865

markstory committed:
Improve description and fix phpcs error

markstory committed:
More phpcs fixes.

markstory committed:
Merge pull request #14932 from cakephp/container-interface

Add Core/ContainerInterface

markstory committed:
Add Application interface and BasePlugin methods

Add the container interface for applications and BasePlugin method.
I've not added the new method to PluginInterface as it could break
existing plugins that don't extend BasePlugin. If this isn't a scenario
we are concerned about I can update the interface.

markstory committed:
Fix phpcs and PHP7.2 typecheck errors.

robertpustulka commented on src/Core/ContainerInterface.php:
```diff
@@ -0,0 +1,55 @@
+     * @param ?bool $shared Set to true to make a service shared.
+     * @return \League\Container\Definition\DefinitionInterface
+     */
+    public function add(string $id, $concrete = null, ?bool $shared = null): DefinitionInterface;
```
Maybe this should be split into three methods each for individual case: `addClassName`, `addObject` and `addClosure`?

markstory replied:
That would diverge from league/container. We might be swapping that library out though as we've not had much luck so far in getting php8 support added.

robertpustulka commented on src/Core/ContainerInterface.php:
```diff
@@ -0,0 +1,55 @@
+     * @param \League\Container\ServiceProvider\ServiceProviderInterface $provider The service provider to add.
+     * @return $this
+     */
+    public function addServiceProvider($provider);
```
Should the parameter be type hinted in method definition?

ADmad replied:
We should probably make our own interface which extends `league/container`'s interface and use that for typehints to be safe. I am vary of directly using external interfaces/classes in case the lib becomes unmaintained in future, like `Aura.Intl` for example.

markstory replied:
I didn't add a typehint as I was trying to keep the interface aligned with what is in league/container. I am happy to revisit the container interface once everything is hooked up and working.

robertpustulka reviewed (commented):
I'm not sure I like sticking to the League implementation of DIC, but this is fine by me as long as the interface is there just for the out-of-the-box plugin compatibility and that I can swap the container in my applications.

markstory committed:
Fix annotation.

markstory committed:
Merge pull request #14946 from cakephp/dic-class

Add Application interface and BasePlugin methods

robertpustulka commented on src/Core/PluginInterface.php:
```diff
@@ -20,6 +20,9 @@
 /**
  * Plugin Interface
+ *
+ * @method \Cake\Core\ContainerInterface register(\Cake\Core\ContainerInterface $container) Register plugin services to
```
Maybe this method should be declared as a part of separate `ContainerPluginInterface`?

markstory replied:
Would we then merge the two interfaces in 5.0? I thought an interface with a single method in it was a bit of a waste.

robertpustulka replied:
I'd like to avoid interface method confusion we had in 3.x with entity methods being added to interface in docblocks.

markstory committed:
Add container support to ControllerFactory

Add the container as a constructor parameter to ControllerFactory. This
makes it possible for controllers to have dependencies injected via
constructor parameters or as parameters to actions.

markstory committed:
Add container methods to BaseApplication

Add the container building methods to BaseApplication and integrate in
the new ControllerFactory constructor change.

markstory committed:
Relax interface requirements.

markstory committed:
Fix missing property.

markstory committed:
Improve dependency resolution for controller actions

Handle passed parameters and optional services better.

markstory committed:
Reduce scope of primitive typehints.

markstory committed:
Fix phpcs errors.

markstory committed:
Save a condition by trimming the ? off.

markstory committed:
Add container support to CommandFactory

This enables commands to have dependencies injected as constructor
arguments.

markstory committed:
Merge pull request #14959 from cakephp/dic-controllerfactory

4.next - Add container to ControllerFactory

markstory committed:
Update annotation.

markstory committed:
Fix another type error.

markstory committed:
Merge branch '4.next' into feature-dic

ADmad committed:
Avoid suppressing error reported by static analyzers.

dereuromark committed:
Merge pull request #14975 from cakephp/reflection-typed-name

Avoid suppressing error reported by static analyzers.

othercorey committed:
Merge pull request #14970 from cakephp/dic-commandfactory

Add container support to CommandFactory

markstory committed:
Add Configuration wrapper.

This wrapper would provide a typehintable interface for configuration
data and the DI container. Typehinting against array configuration is
not possible with most container implementations. If folks are happy
with this approach a Configuration instance can be added to the default
container as a shared service.

markstory committed:
Fix phpcs

markstory committed:
Rename injectable config proxy.

markstory committed:
Fix import order.

markstory committed:
Merge pull request #14977 from cakephp/dic-configuration

Add Configuration wrapper.

markstory committed:
Add docs for runCommand + container

Add notes on the limitations of runCommand() and the container.
Because commands don't have access to the container they are unable
resolve dependencies. Instead commands should be defined in the
container and then resolved as dependencies to the calling command.

markstory committed:
Merge pull request #14982 from cakephp/dic-runcommand

Add docs for runCommand + container

markstory commented:
I think this has all of what I outlined in the linked issue. One potential addition I could see is an event for the container being built, but I don't have a good use case for it just yet.

markstory commented:
Any other feedback on this?

othercorey commented:
Other than the `register()` method as an annotation on `PluginInterface` looks good to merge for now. I don't know that we really have a better solution than annotations until cakephp 5 is released. We can consider containers an experimental feature until then probably.

markstory commented:
> Other than the register() method as an annotation on PluginInterface looks good to merge for now. I don't know that we really have a better solution than annotations until cakephp 5 is released. 

Would you prefer it be a concrete method on PluginInterface?

othercorey commented:
> Would you prefer it be a concrete method on PluginInterface?

It should be, but since we're adding this mid-release to an interface it's what it has to be. We definitely don't want to hide the container api for cakephp 5.

I'm not personally a fan of just calling it `register`, but I leave that up to you.

robertpustulka commented:
> Would you prefer it be a concrete method on PluginInterface?

It can be a new interface (something like `ContainerPluginInterface` similar to `ContainerApplicationInterface`) that the base Plugin can implement.

markstory commented:
> I'm not personally a fan of just calling it register, but I leave that up to you.

I'm open to changing the names of the methods if we have alternatives in mind.

ADmad commented:
`register()` is inline with naming used in `league/container`. There's `ServiceProviderInterface::register()` https://container.thephpleague.com/3.x/service-providers/

othercorey commented:
I was just thinking in the context of a Plugin - register is a little generic. It doesn't really imply container-specific stuff.
